src/algebra/group/freiman.lean

```diff
@@ -0,0 +1,337 @@
+/-
+Copyright (c) 2021 Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies
+-/
+import data.fun_like
+import data.multiset.basic
+
+/-!
+# Freiman homomorphisms
+
+In this file, we define Freiman homomorphisms. A `n`-Freiman homomorphism is a function `f : α → β`
+such that `f (x₁) * ... * f (xₙ) = f (y₁) * ... * f (yₙ)` for all `x₁, ..., xₙ, y₁, ..., yₙ` such
+that `x₁ * ... * xₙ = y₁ * ... * yₙ`. In particular, any `mul_hom` is a Freiman homomorphism.
+
+They are of interest in additive combinatorics.
+
+## Main declaration
+
+* `freiman_hom`: Freiman homomorphism.
+* `add_freiman_hom`: Additive Freiman homomorphism.
+
+## Notation
+
+* `α →*[n] β`: Multiplicative `n`-Freiman homomorphism
+* `α →+[n] β`: Additive`n`-Freiman homomorphism
+
+## TODO
+
+`monoid_hom.to_freiman_hom` could be relaxed to `mul_hom.to_freiman_hom` by proving
+`(s.map f).prod = (t.map f).prod` directly by induction instead of going through `f s.prod`.
+-/
+
+open multiset
+
+variables {F α β γ δ G : Type*}
+
+/-- An additive `n`-Freiman homomorphism is a map which preserves sums of `n` elements. -/
+structure add_freiman_hom (α β : Type*) [add_comm_monoid α] [add_comm_monoid β] (n : ℕ) :=
+(to_fun : α → β)
+(map_sum_eq_map_sum' {s t : multiset α} (hs : s.card = n) (ht : t.card = n) (h : s.sum = t.sum) :
+  (s.map to_fun).sum = (t.map to_fun).sum)
+
+/-- A `n`-Freiman homomorphism is a map which preserves products of `n` elements. -/
+@[to_additive add_freiman_hom]
+structure freiman_hom (α β : Type*) [comm_monoid α] [comm_monoid β] (n : ℕ) :=
+(to_fun : α → β)
+(map_prod_eq_map_prod' {s t : multiset α} (hs : s.card = n) (ht : t.card = n)
+  (h : s.prod = t.prod) : (s.map to_fun).prod = (t.map to_fun).prod)
+
+notation α ` →+[`:25 n:25 `] `:0 β:0 := add_freiman_hom α β n
+notation α ` →*[`:25 n:25 `] `:0 β:0 := freiman_hom α β n
+
+/-- `add_freiman_hom_class F α β n` states that `F` is a type of `n`-ary sums-preserving morphisms.
+You should extend this class when you extend `add_freiman_hom`. -/
+class add_freiman_hom_class (F : Type*) (α β : out_param $ Type*) [add_comm_monoid α]
+  [add_comm_monoid β] (n : ℕ) extends fun_like F α (λ _, β) :=
+(map_sum_eq_map_sum' (f : F) {s t : multiset α} (hs : s.card = n) (ht : t.card = n)
+  (h : s.sum = t.sum) : (s.map f).sum = (t.map f).sum)
+
+/-- `freiman_hom_class F α β n` states that `F` is a type of `n`-ary products-preserving morphisms.
+You should extend this class when you extend `freiman_hom`. -/
+@[to_additive add_freiman_hom_class]
+class freiman_hom_class (F : Type*) (α β : out_param $ Type*) [comm_monoid α] [comm_monoid β]
+  (n : out_param ℕ)
```

Comment:
Both `n`s should be `out_param`s here, otherwise the `to_fun_like` projection won't work.
```suggestion
  [add_comm_monoid β] (n : out_param ℕ) extends fun_like F α (λ _, β) :=
(map_sum_eq_map_sum' (f : F) {s t : multiset α} (hs : s.card = n) (ht : t.card = n)
  (h : s.sum = t.sum) : (s.map f).sum = (t.map f).sum)

/-- `freiman_hom_class F α β n` states that `F` is a type of `n`-ary products-preserving morphisms.
You should extend this class when you extend `freiman_hom`. -/
@[to_additive add_freiman_hom_class]
class freiman_hom_class (F : Type*) (α β : out_param $ Type*) [comm_monoid α] [comm_monoid β]
  (n : out_param ℕ)
```

Reply:
Based on our conversation on Zulip, these `n`s can't be `out_param`s. So we'll have to unbundle the `fun_like` part...